tenpy/algorithms/tebd.py

```diff
@@ -591,6 +591,264 @@ def __init__(self, psi, model, options):
+        #  theta_new = npc.tensordot(npc.tensordot(A_L, Xi, ['vR', 'vL']), B_R.replace_label('p', 'p1'), ['vR', 'vL'])
+        #  print("<theta | A Xi B> =", npc.inner(theta, theta_new, axes='labels', do_conj=True))
+        return A_L, Xi, B_R
+
+    def update_bond_imag(self, i, U_bond):
+        i0, i1 = i - 1, i
+        expand = self._expansion_rate(i)
+        logger.debug(f'Update sites ({i0}, {i1}). CBE expand={expand}')
+        # Construct the theta matrix
+        theta = self.psi.get_theta(i0, n=2)
+        theta = npc.tensordot(U_bond, theta, axes=(['p0*', 'p1*'], ['p0', 'p1']))
+        theta.itranspose(['vL', 'p0', 'p1', 'vR'])
+
+        A_L, Xi, B_R = self._qr_based_svd(i, theta, expand)
+        Xi.itranspose(['vL', 'vR'])
+
+        if self.options.get('use_eig_based_svd', False):
+            U, S, Vd, trunc_err, renormalize = _eig_based_svd(
+                Xi, inner_labels=['vR', 'vL'], trunc_params=self.trunc_params
+            )
+        else:
+            U, S, Vd, trunc_err, renormalize = svd_theta(Xi, self.trunc_params)
+
+        B_R = npc.tensordot(Vd, B_R, axes=['vR', 'vL'])
+        A_L = npc.tensordot(A_L, U, axes=['vR', 'vL'])
+
+        if self.options.get('compute_err', True):
+            trunc_err = TruncationError.from_norm(norm_new=np.linalg.norm(S), norm_old=npc.norm(theta))
+        else:
+            trunc_err = TruncationError(np.nan, np.nan)
+
+        self.psi.norm *= renormalize
+        self.psi.set_B(i0, A_L, form='A')
+        self.psi.set_SL(i1, S)
+        self.psi.set_B(i1, B_R, form='B')
+
+        return trunc_err
+
+
+def _eig_based_svd(A, need_U: bool = True, need_Vd: bool = True, inner_labels=[None, None],
+                   trunc_params=None):
+    """Computes the singular value decomposition of a matrix A, but via diagonalization of
+    its "square" A.hc @ A and/or A @ A.hc, i.e. two eigh calls instead of an svd call.
+
+    Truncation if performed if and only if trunc_params are given.
+    This performs better on GPU, but is not really useful on CPU.
+    If isometries U or Vd are not needed, their computation can be omitted for performance.
+    """
+    warnings.warn('_eig_based_svd is nonsensical on CPU!!')
+    assert A.rank == 2
+
+    if need_U and need_Vd:
+        # need to do two diagonalizations to get both right and left singular vectors
+        try:
+            A.legs[0].test_contractible(A.legs[1])
+        except ValueError:
+            # need to be more elaborate in this case ; the new leg on U as computed below
+            # would not be contractible with the new leg in Vd ...
+            raise NotImplementedError from None
+        
+        U, S, _, _, _ = _eig_based_svd(
+            A, need_U=True, need_Vd=False, inner_labels=inner_labels, trunc_params=trunc_params
+        )
+        _, S, Vd, trunc_err, renormalize = _eig_based_svd(
+            A, need_U=False, need_Vd=True, inner_labels=inner_labels, trunc_params=trunc_params
+        )
```

Comment:
I don't think that's completely correct, we need to get Vd from U and S.
At least phases of U/Vd are arbitrary, such that `M != U S Vd` in most cases.

There might also be a danger that S from the different sides are not in corresponding order in case of degenerate singular values + machine precision noise? But I'm not sure if this is an issue (beyond the phases), since the singular vectors themselves are not uniquely defined in that case, you just need any ONB of that subspace...